### ISSUE TITLE
How to use Flops Profiler with Pipeline Parallelism?

### ISSUE THREAD

**Issue by flyingdown** [microsoft/Megatron-DeepSpeed#129]
if I train a model with pipeline parallerlism, Flops Profiler in ds config is useless. Add some code like this:

prof = None
    if args.deepspeed:
        prof = FlopsProfiler(model[0])
    else:
        prof = FlopsProfiler(model)
...
train_step(forward_step_func,
                       train_data_iterator,
                       model,
                       optimizer,
                       lr_scheduler)
....
if iteration == profile_step and mpu.get_data_parallel_rank() == 0 and mpu.is_pipeline_last_stage() and mpu.get_tensor_model_parallel_rank() == 0:
            prof.print_model_profile(profile_step=profile_step)
            prof.end_profile()

but output only has fwd info, like:

-------------------------- DeepSpeed Flops Profiler --------------------------
Profile Summary at step 8:
Notations:
data parallel size (dp_size), model parallel size(mp_size),
number of parameters (params), number of multiply-accumulate operations(MACs),
number of floating-point operations (flops), floating-point operations per second (FLOPS),
fwd latency (forward propagation latency), bwd latency (backward propagation latency),
step (weights update latency), iter latency (sum of fwd, bwd and step latency)

params per gpu:                                               463.21 M
params of model = params per GPU * mp_size:                   463.21 M
fwd MACs per GPU:                                             242304.87 GMACs
fwd flops per GPU:                                            484678.47 G
fwd flops of model = fwd flops per GPU * mp_size:             484678.47 G
fwd latency:                                                  29.87 s
fwd FLOPS per GPU = fwd flops per GPU / fwd latency:          16.22 TFLOPS

----------------------------- Aggregated Profile per GPU -----------------------------
Top 1 modules in terms of params, MACs or fwd latency at different model depths:
depth 0:
    params      - {'PipelineEngine': '463.21 M'}
    MACs        - {'PipelineEngine': '242304.87 GMACs'}
    fwd latency - {'PipelineEngine': '29.87 s'}
depth 1:
    params      - {'GPTModelPipe': '463.21 M'}
    MACs        - {'GPTModelPipe': '242304.87 GMACs'}
    fwd latency - {'GPTModelPipe': '29.87 s'}
depth 2:
    params      - {'ParallelTransformerLayerPipe': '402.91 M'}
    MACs        - {'ParallelTransformerLayerPipe': '228698.42 GMACs'}
    fwd latency - {'ParallelTransformerLayerPipe': '55.83 s'}
depth 3:
    params      - {'ParallelMLP': '268.5 M'}
    MACs        - {'ParallelMLP': '140737.49 GMACs'}
    fwd latency - {'ParallelMLP': '27.79 s'}

------------------------------ Detailed Profile per GPU ------------------------------
Each module profile is listed after its name in the following order:
params, percentage of total params, MACs, percentage of total MACs, fwd latency, percentage of total fwd latency, fwd FLOPS

Note: 1. A module can have torch.nn.module or torch.nn.functional to compute logits (e.g. CrossEntropyLoss). They are not counted as submodules, thus not to be printed out. However they make up the difference between a parent's MACs (or latency) and the sum of its submodules'.
2. Number of floating-point operations is a theoretical estimation, thus FLOPS computed using that could be larger than the maximum system throughput.
3. The fwd latency listed in the top module's profile is directly captured at the module forward function in PyTorch, thus it's less than the fwd latency shown above which is captured in DeepSpeed.

 iteration        9/   29551 | consumed samples:         1152 | consumed tokens:      2359296 | elapsed time per iteration (ms): 112211.1 | learning rate: 6.000E-05 | global batch size:   128 | lm loss: 1.275906E+01 | loss scale: 4096.0 | grad norm: 8.305 | num zeros: 0.0 | actual seqlen:  2048 | number of skipped iterations:   0 | number of nan iterations:   0 | samples per second: 1.141 | TFLOPs: 8.33 |
time (ms)
PipelineEngine(
  463.21 M, 100.00% Params, 242304.87 GMACs, 100.00% MACs, 29.87 s, 100.00% latency, 16.22 TFLOPS,

**Comment by forrestjgq** [comment]
same issue here

**Comment by Jiminator** [comment]
Have anyone ever managed to solve this issue? Also trying to use the FLOPS profiler with Pipeline Parallelism, with no luck
